kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt

```diff
@@ -129,9 +135,12 @@ class TunnelService : VpnService() {
             }
 
             override fun onDisconnect(error: String?): Boolean {
-                Log.d(TAG, "onDisconnect $error")
+                Log.d(TAG, "callback: onDisconnect $error")
+
+                if (error != null && error != "null") {
```

Comment:
If `error` is null here, where does the `sessionPtr` and tunnel get cleaned up? Maybe I'm not following, but doesn't it make sense to always clean up the Android-side of things here?

From my understanding, a user-initiated disconnect is the same flow as a connlib-initiated one, the only differences being:

- User-initiated entrypoint is `TunnelSession.disconnect(it)`
- `onDisconnect(error = null)`

I don't think connlib does any extra cleanup if the onDisconnect is user-initiated or other scenarios.
cc @conectado

Reply:
Yes, that's correct. there are two disconnect paths:

- User initiated.
1. initiated by `TunnelSession.disconnect(long)` (TunnelService.kt:204)
2. connlib disconnects and the `callback.onDisconnect(null)` is called
3. the flow continues to call the `TunnelService.onDisconnect` where the pointers and listeners are cleared.

- On connlib error
1. In case of non-recoverable errors, `callback.pmDisconnect(error)` is called, with non-null error parameter.
2. here the `TunnelService.onDisconnect` gets called, which clears the pointers and listeners.

- Also I added checks in the service class to handle cases when the app is restarted.

Reply:
> the flow continues to call the TunnelService.onDisconnect where the pointers and listeners are cleared.

That's my point though -- doesn't this conditional prevent the pointers and listeners from being cleared in the user-initiated case?

```
if (error != null && error != "null") {
    onDisconnect()
}
```

Reply:
I am skipping the `onDisconnect` call in the callback as during the user-initiated flow, from the `disconnect()` function (TunnelService.kt:204), in this function it calls the `TunnelSession.disconnect`  and then ultimately calls `onDisconnect` function in the end which clears the pointer and listeners.

```kotlin
private fun disconnect() {
        Log.d(TAG, "disconnect(): Attempting to disconnect session")
        try {
            sessionPtr?.let {
                TunnelSession.disconnect(it)
            }
        } catch (exception: Exception) {
            Log.e(TAG, exception.message.toString())
        }

        onTunnelStateUpdate(Tunnel.State.Down)

        if (shouldReconnect) {
            shouldReconnect = false
            connect()
        } else {
            onDisconnect()
        }
}
```

Reply:
My thoughts here are:

I'm unclear when in `disconnect` is `shouldReconnect` true? 

Also, since this callback will always be called after the tunnel is down, can we just let this callback always call `onDisconnect` and never call it from `disconnect`?

Reply:
Agree with Gabi here -- the logic here seems confusing. I would have thought `onDisconnect` behaves the same regardless of its `error` parameter value.

Reply:
@jamilbk @conectado I have refactored the implementation, please check.